src/cargo/sources/path.rs

```diff
@@ -162,50 +175,52 @@ impl<'cfg> PathSource<'cfg> {
-        filter: &mut dyn FnMut(&Path) -> CargoResult<bool>,
-    ) -> Option<CargoResult<Vec<PathBuf>>> {
-        // If this package is in a Git repository, then we really do want to
-        // query the Git repository as it takes into account items such as
-        // `.gitignore`. We're not quite sure where the Git repository is,
-        // however, so we do a bit of a probe.
-        //
-        // We walk this package's path upwards and look for a sibling
-        // `Cargo.toml` and `.git` directory. If we find one then we assume that
-        // we're part of that repository.
-        let mut cur = root;
-        loop {
-            if cur.join("Cargo.toml").is_file() {
-                // If we find a Git repository next to this `Cargo.toml`, we still
-                // check to see if we are indeed part of the index. If not, then
-                // this is likely an unrelated Git repo, so keep going.
-                if let Ok(repo) = git2::Repository::open(cur) {
-                    let index = match repo.index() {
-                        Ok(index) => index,
-                        Err(err) => return Some(Err(err.into())),
-                    };
-                    let path = root.strip_prefix(cur).unwrap().join("Cargo.toml");
-                    if index.get_path(&path, 0).is_some() {
-                        return Some(self.list_files_git(pkg, &repo, filter));
-                    }
-                }
-            }
-            // Don't cross submodule boundaries.
-            if cur.join(".git").is_dir() {
-                break;
-            }
-            match cur.parent() {
-                Some(parent) => cur = parent,
-                None => break,
+        filter: &mut dyn FnMut(&Path, bool) -> CargoResult<bool>,
+    ) -> CargoResult<Option<Vec<PathBuf>>> {
+        let repo = match git2::Repository::discover(root) {
+            Ok(repo) => repo,
+            Err(e) => {
+                log::debug!(
+                    "could not discover git repo at or above {}: {}",
+                    root.display(),
+                    e
+                );
+                return Ok(None);
             }
+        };
+        let index = repo
+            .index()
+            .chain_err(|| format!("failed to open git index at {}", repo.path().display()))?;
+        let repo_root = repo.workdir().ok_or_else(|| {
+            anyhow::format_err!(
+                "did not expect repo at {} to be bare",
+                repo.path().display()
+            )
+        })?;
+        let repo_relative_path = root.strip_prefix(repo_root).chain_err(|| {
+            format!(
+                "expected git repo {} to be parent of package {}",
+                repo.path().display(),
+                root.display()
+            )
+        })?;
+        // Git requires forward-slashes.
+        let repo_safe_path = repo_relative_path
+            .join("Cargo.toml")
+            .to_string_lossy()
+            .replace('\\', "/");
```

Comment:
I have submitted https://github.com/rust-lang/git2-rs/pull/549 as an upstream fix for this.  I think it might be good to wait to merge that, then this can be removed.

Reply:
@alexcrichton would you mind publishing a new git2 release, and then I can remove this?

Reply:
er oops, done now!